File: sprint3/By languages/C++/lesson11code1.cpp
#include <iostream>
#include <vector>

using namespace std;

vector<int> merge_sort(vector<int>& array) {
    if (array.size() == 1) {  // базовый случай рекурсии
        return array;
    }

    // запускаем сортировку рекурсивно на левой половине
    vector<int> left(array.begin(), array.begin() + array.size() / 2);
    left = merge_sort(left);

    // запускаем сортировку рекурсивно на правой половине
    vector<int> right(array.begin() + array.size() / 2, array.end());
    right = merge_sort(right);

    // заводим массив для результата сортировки
    vector<int> result(array.size());

    // сливаем результаты
    int l = 0, r = 0, k = 0;
    while (l < left.size() && r < right.size()) { 
    // выбираем, из какого массива забрать минимальный элемент
        if (left[l] <= right[r]) { 
            result[k] = left[l];
            l++;
        }
        else {
            result[k] = right[r];
            r++;
        }
        k++;
    }

    // Если один массив закончился раньше, чем второй, то
    // переносим оставшиеся элементы второго массива в результирующий
    while (l < left.size()) { 
        result[k] = left[l];   // перенеси оставшиеся элементы left в result
        l++;
        k++;
    }  
    while (r < right.size()) { 
        result[k] = right[r];  // перенеси оставшиеся элементы right в result
        r++;
        k++;
    }

    return result;
}

int main() {
    vector<int> arr = {5, 2, 7, 1, 8, 3};
    arr = merge_sort(arr);
    for (int i = 0; i < arr.size(); i++) {
        cout << arr[i] << " ";
    }
    cout << endl;
    return 0;
}

File: sprint3/By languages/go/lesson11code1.go
func mergeSort(array []int) []int {
    if len(array) == 1 {  // базовый случай рекурсии
        return array
    }

    // запускаем сортировку рекурсивно на левой половине
    left := mergeSort(array[0:len(array)/2])

    // запускаем сортировку рекурсивно на правой половине
    right := mergeSort(array[len(array)/2:len(array)])

    // заводим массив для результата сортировки
    result := make([]int, len(array))

    // сливаем результаты
    l, r, k := 0, 0, 0
    for l < len(left) && r < len(right) {
        // выбираем, из какого массива забрать минимальный элемент
        if left[l] <= right[r] {
            result[k] = left[l]
            l++
        } else {
            result[k] = right[r]
            r++
        }
        k++
    }

    // Если один массив закончился раньше, чем второй, то
    // переносим оставшиеся элементы второго массива в результирующий
    for l < len(left) {
        result[k] = left[l]  // перенеси оставшиеся элементы left в result
        l++
        k++
    }
    for r < len(right) {
        result[k] = right[r] // перенеси оставшиеся элементы right в result
        r++
        k++
    }

    return result
}
